spec/light-client/Lightclient.tla

```diff
@@ -1,111 +1,273 @@
+    /\ LET initStack == << [isLeft |-> TRUE,
+                           startHeight |-> TRUSTED_HEIGHT,
+                           endHeight |-> inEvent.heightToVerify] >>
+       IN           
+        /\ requestStack' = initStack
+        /\ outEvent' = RequestHeaderForTopRequest(initStack)
+
+(* Check whether we can trust the signedHdr based on trustedHdr
+   following the trusting period method.
+   This operator is similar to CheckSupport in the English spec.
+   
+   The parameters have the following meanings:
+   - heightToTrust is the height of the trusted header
+   - heightToVerify is the height of the header to be verified
+   - trustedHdr is the trusted header (not a signed header)
+   - signedHdr is the signed header to verify (including commits)
+   *)
+CheckSupport(heightToTrust, heightToVerify, trustedHdr, signedHdr) ==
+    IF minTrustedHeight > heightToTrust \* outside of the trusting period
+        \/ ~(signedHdr[2] \subseteq (DOMAIN signedHdr[1].VP)) \* signed by other nodes
+    THEN FALSE
+    ELSE
+      LET TP == BC!PowerOfSet(trustedHdr.NextVP, DOMAIN trustedHdr.NextVP)
+          SP == BC!PowerOfSet(trustedHdr.NextVP,
+            signedHdr[2] \intersect DOMAIN trustedHdr.NextVP)
+      IN
+      IF heightToVerify = heightToTrust + 1
+          \* the special case of adjacent heights: check 2/3s
+      THEN (3 * SP > 2 * TP)
+      ELSE \* the general case: check 1/3 between the headers and make sure there are 2/3 votes  
+        LET TPV == BC!PowerOfSet(signedHdr[1].VP, DOMAIN signedHdr[1].VP)
+            SPV == BC!PowerOfSet(signedHdr[1].VP, signedHdr[2])
+        IN
+        (3 * SP > TP) /\ (3 * SPV > 2 * TPV)
+
+(* Make one step of bisection, roughly one stack frame of Bisection in the English spec *)
+OneStepOfBisection(storedHdrs) ==
+    LET topReq == Head(requestStack)
+        lh == topReq.startHeight
+        rh == topReq.endHeight
+        lhdr == CHOOSE hdr \in storedHdrs: hdr[1].height = lh
+        rhdr == CHOOSE hdr \in storedHdrs: hdr[1].height = rh
+    IN
+    \* pass only the header lhdr[1] and signed header rhdr
+    IF CheckSupport(lh, rh, lhdr[1], rhdr)
+        (* The header can be trusted, pop the request and return true *)
+    THEN <<TRUE, Tail(requestStack)>>
+    ELSE IF lh + 1 = rh \* sequential verification
+        THEN (*
+             Sequential verification tells us that the header cannot be trusted:
+             If the search request was scheduled as a left branch, then
+             (1) pop the top request (the left branch), and
+             (2) pop the second top request, which is the request for the right branch.
+             Otherwise, pop only the top request.
+             This is the optimization that is introduced in the English spec.
+             One can replace it e.g. with a parallel search.
+             *)
+            LET howManyToPop == IF topReq.isLeft THEN 2 ELSE 1 IN  
+            <<FALSE, SubSeq(requestStack, 1 + howManyToPop, Len(requestStack))>>
```

Comment:
First I think an explanation on what SubSeq would be useful:
`SubSeq(s, m, n) == [i \in 1..(1+n-m) |-> s[i+m-1]]
  (* The sequence <<s[m], s[m+1], ... , s[n]>>.                            *)
`
If we start with an `initStack` like: `<<(T, N, N+1)>>` and we are at 181 with failed `CheckSupport` then:
- 181: `howManyToPop` == 2
- 182: will call `SubSeq(initStack, 3, 1)` -> I assume this will just return empty sequence. Just a bit awkward.

But another potential problem is that in 182 we may have non-empty stack which takes the caller to line 206 where the verdict is not checked and we continue verification of right branch with a failure in left. In general I don't understand this optimization. Which english spec you refer to here?

Reply:
We used the [spec](https://github.com/tendermint/tendermint/blob/master/docs/spec/consensus/light-client.md) as the input. Essentially, when one pops two elements, the top-level Bisection call returned false and thus the inner call to Bisection should never occur. This looks tricky, as we have to encode two non-tail recursive calls using the stack (Bisection in [English spec](https://github.com/tendermint/tendermint/blob/master/docs/spec/consensus/light-client.md) cannot be easily converted into a loop).

Reply:
Yes but the bisection should stop only if >2/3rds has not signed. In the other cases it should continue.

Reply:
Fixed. Now it is much simpler. See b06216bb54f8d26a8b1d6cc0123546dcbcb18957